rfc/006-semantic-versioning.md

```diff
@@ -0,0 +1,76 @@
+
+This document attempts to encapsulate the discussions around versioning in Tendermint and draws upon them to propose a guide to how Tendermint uses versioning to make promises to its users.
+
+It's worth to further acknowledge that the strictness of our versioning guarantees may mean nothing if we don't also address the frequency that the team intends on making version changes. We could make the strictest guarantees in the world yet simply break them with every release.
+
+Finally I would like to remark that this RFC only addresses the what, as in what are the rules for versioning. The how, or how does Tendermint plan to follow the versioning logic, will be addressed in a later RFC on Soft Upgrades.
+
+## Discussion
+
+We first begin with a round up of the various users and a set of assumptions on what these users expect from Tendermint in regards to versioning:
+
+1. Application Developers, those that use the ABCI to build applications on top of Tendermint, are chiefly concerned with that API. Breaking changes will force developers to modify large portions of their codebase to accommodate for the changes. Some ABCI changes such as introducing priority for the mempool don't require any effort and can be lazily adopted whilst changes like ABCI++ may force applications to redesign their entire execution system.
+2. Node Operators, those running node infrastructure, are predominantly concerned with downtime, complexity and frequency of upgrading and loss of data. They may be also concerned about changes that may break the scripts and tooling they use to supervise their nodes.
+3. External clients, for example, wallets and block explorers are those that perform any of the following:
+     - consume the RPC endpoints of nodes like `/block`
+     - subscribe to events that are streamed via websockets
+     - make queries to the indexer
+  
+    This set are concerned with chain upgrades which will impact their ability to query state and block data as well as broadcast transactions.
+  
+4. There is also a fourth emerging user: IBC module and relayers (although in many ways relayers can also be considered as external clients). The developers of IBC and consumers of their software are concerned about changes that may affect a chains ability to send arbitrary messages to another chain. Specifically, this can happen with a breaking change to the light client verification algorithm.  
```

Comment:
Thank you for acknowledging IBC mod & relayers as a class of users that are affected by changes in tm-go! It seems that this category of users is in some sense a superset of (3). It's not clear to me, however, what part of tm-go affects the light client verification algorithm: is it any changes to block headers, for example?

Reply:
Exactly. Part of the feature set for 0.36 includes modifications to the `CommitSig` and `Commit`. The questions I have been asking the ibc-go team is that with such changes, could a chain that doesn't upgrade their IBC module still be able to verify packets (using the light client algorithm) with a chain that had upgraded. If not, then this Tendermint change effectively breaks IBC and chains across different versions won't be able to communicate. Changes to `ValidatorSet`, `Commit` or `Header` could all potentially break IBC.

Reply:
I think the distinguishing feature of this set of clients is cross-chain interoperability. Perhaps we can find a better name, but other clients do not need this type of interoperability, they just need to be able to parse out the relevant data structures. Clients performing cross-chain operations are in a special category of potentially requiring both chains to match completely, or nearly completely, which isn't a requirement of other applications like wallets etc.

Reply:
Hmm. So maybe "multi-chain external clients"?

Reply:
I was talking about something similar to this with @creachadair yesterday. My interpretations is that there are two types of clients to a blockchain: one that queries information from a trusted source and the other that wants to validate information. I'd ideally like to map this distinction to our two IPC layers: P2P for verification and RPC for information hence the reason for wanting to bring the light client into the p2p stack - that way clients that want verification run their light client in the browser and query it for information via RPC. The only tricky situation in all of this is the IBC module which uses the RPC layer and is outside of what we'd consider the network boundary yet requires data for verification purposes. I will write some notes up.